### PR TITLE
Enhancement: Generate cookie cutter templates and apply linting on them

### LINKED ISSUE #122: Add ansible-lint for cookiecutter in CI jobs
Related to #119

Currently, the cookiecutter directory is excluded in ansible-lint configuration.

It may need to run 'molecule init -d xxx' to generate the playbook from cookiecutter.

The lint issue in Azure plugin has been fix in #105

### PR DESCRIPTION
**Pull request by wadhah101**
## closes #122

### CODE REVIEW

File: tools/generate-templates.sh
Comment (zhan9san) [inline tools/generate-templates.sh:5]: Could it be renamed to `test/roles`?

It's a coincidence. In `ansible-lint`, the file under a directory named `templates` would be considered as a text format instead yaml format. So when `ansible-lint` runs, it would always show no Violations. 

https://github.com/ansible/ansible-lint/blob/aaab27952352edf3aa8afaffa4fc1e93ff984976/src/ansiblelint/config.py#L82

File: tox.ini
Comment (zhan9san) [inline tox.ini:92]: No need to add this line, I think.

It is defined in `[testenv]` already.

File: .ansible-lint
Comment (zhan9san) [inline .ansible-lint:6]: If the `ansible-lint` is added, CI jobs would fail until ansible-lint are happy with playbooks in all plugins.
But now, only azure plugin fixed some lint issues. We may need to exclude playbooks generated from other plugins. e.g. `test/roles/xxx`.

File: tools/generate-templates.sh
Comment (zhan9san) [inline tools/generate-templates.sh:6]: If we create some temporary files, it's better to add them in `.gitignore` as well.

As we may run `tox -e lint` locally, it would generates this temporary files.

File: .ansible-lint
Comment (zhan9san) [inline .ansible-lint:9]: It would be better if the skip_list can be deleted in this PR.

**Review by zhan9san — CHANGES_REQUESTED** [review]

File: requirements.yml
Comment (zhan9san) [inline requirements.yml:12]: Here is an explicit definition.  How about using 1.8.0 here?

https://github.com/ansible-community/molecule-plugins/blob/5ebdfc60e457d5303e45fb237840e6592b98715f/src/molecule_plugins/gce/driver.py#L174-L176

File: .pre-commit-config.yaml
Comment (zhan9san) [inline .pre-commit-config.yaml:43]: Why do we need to define ANSIBLE_ROLES_PATH here?

File: .ansible-lint-ignore
Comment (zhan9san) [inline .ansible-lint-ignore:83]: Here is a long list. It's fun for us to fix the lint issue in another PR and finally, remove the related lines in this files in  new PR.

File: tools/generate-templates.sh
Comment (zhan9san) [inline tools/generate-templates.sh:6]: Every time we run `tox -e lint`, it will generate some temporary files locally, can we add `test/roles` in `.gitignore`?

**Review by zhan9san — CHANGES_REQUESTED** [review]

File: tools/generate-templates.sh
Comment (wadhah101) [reply tools/generate-templates.sh]: DONE

File: tox.ini
Comment (wadhah101) [reply tox.ini]: DONE

File: .ansible-lint
Comment (wadhah101) [reply .ansible-lint:6]: I have added ansible-ignore file instead. I think it's more explicit and less prone to letting new errors slip

File: tools/generate-templates.sh
Comment (wadhah101) [reply tools/generate-templates.sh]: We can't put it into .gitignore because pre-commit will ignore it in linting

File: .ansible-lint
Comment (wadhah101) [reply .ansible-lint]: Done

File: .pre-commit-config.yaml
Comment (wadhah101) [reply .pre-commit-config.yaml:43]: If you we don't define roles path ansbile lint with fail with plugin not found error. couldn't figure out how to solve it 

Example:
`syntax-check[specific]: the role 'ec2plugin' was not found in /Users/wadhahmahroug/Projects/opensource/molecule-plugins/test/roles/ec2plugin/tests/roles:/Users/wadhahmahroug/.cache/ansible-compat/5f579b/roles:/Users/wadhahmahroug/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles:/Users/wadhahmahroug/Projects/opensource/molecule-plugins/test/roles/ec2plugin/tests
test/roles/ec2plugin/tests/test.yml:5:7`

File: .ansible-lint-ignore
Comment (wadhah101) [reply .ansible-lint-ignore:83]: I think in this PR we should leave this list of explicitely ignored errors then solve them in another PR

 FQDN and line length, i will do them in this PR, they are easy to solve

File: tools/generate-templates.sh
Comment (wadhah101) [reply tools/generate-templates.sh:6]: pre-commit will ignore stuff inside .gitignore

File: requirements.yml
Comment (wadhah101) [reply requirements.yml]: Done

File: tox.ini
Comment (zhan9san) [inline tox.ini:93]:
```suggestion
    pre-commit run -a
    rm -rf test/roles
```

Comment (zhan9san) [reply tox.ini:93]: You are right. The .gitignore doesn't work as expect.

Let us clean up the temporary files explicitly at the end of test.

**Comment by wadhah101** [comment]
I will fix the remaining errors later and have a final commit for this PR @zhan9san

**Comment by zhan9san** [comment]
@wadhah101 

There are much many lint issues now. It's tedious to add them one by one in `.ansible-lint-ignore`.

We can add `exclude_paths` to exclude the whole folders.

In this PR, it is fine to enabl ansible-lint in CI only. Try to make the PR as smaller.

**Comment by apatard** [comment]
Can you please clean this PR a little bit ?
Current state doesn't make it easy to review (at least to me).

Moreover, I see changes which should probably need their own PR:
- azure VM size and default VM image (as said in PR #105),
- lint issues with playbooks like the docker ones or with the test scenarios

I'm going to add also some quick comments to some of the changes done. There will possibly more later, one this PR is clean.

File: .ansible-lint
Comment (apatard) [inline .ansible-lint:11]: Please don't remove args[module] from skip_list. I can't comment for @ssbarnea but in its current form, this rule is kind of broken, at least with modules relying on an action plugin.

File: .ansible-lint-ignore
Comment (apatard) [reply .ansible-lint-ignore:51]: I guess that meta errors are going to stay, unless the test script modifies the generated meta.yml file. Would have been easier if ansible-lint could support wildcards...

File: .pre-commit-config.yaml
Comment (apatard) [reply .pre-commit-config.yaml:43]: did you try changing to the FQDN and add the collection providing it in the requirements.yml file ?

File: tools/generate-templates.sh
Comment (apatard) [reply tools/generate-templates.sh:6]: I may be missing something but why not creating a temporary directory with mktemp ? issue with non linux os like macosx ?

File: tools/generate-templates.sh
Comment (apatard) [inline tools/generate-templates.sh:9]: Is there a way to not hardcode the list ? Each time a new plugin is added, we'll have to edit this script. This can be easily forgotten

File: tools/generate-templates.sh
Comment (zhan9san) [reply tools/generate-templates.sh:6]: What benefits`mktemp` can bring?

IMHO, `mktemp` is better for `unique` file name. In this PR, we just want to create a directory and generate some files in them.

Comment (apatard) [reply tools/generate-templates.sh:6]: I find it cleaner. It won't do any change in the working directory and it should avoid having to deal with configurations files of tools like git or pre-commit to ignore it.

File: .ansible-lint
Comment (wadhah101) [reply .ansible-lint:11]: Restored it 
Done

File: tools/generate-templates.sh
Comment (wadhah101) [reply tools/generate-templates.sh:9]: I can list the folders inside `src/molecule_plugins` but do you think this a good idea ?.

I fear that other folder might be added that are cookiecutter templates

**Comment by wadhah101** [comment]
@apatard I have cleaned up the code and only left changes relevant to making the ci functional.

Any syntax fixes for ansbile I will do in other PRs.

About VM size I plan to open another PR for it but no this one.

File: .pre-commit-config.yaml
Comment (apatard) [reply .pre-commit-config.yaml:43]: Please ignore my comment, I misread the error message.

File: tools/generate-templates.sh
Comment (apatard) [reply tools/generate-templates.sh:6]: hm. maybe it would be easier to fix all errors before merging this PR, _maybe_ this would make the .ansible-lint-ignore file useless ?

File: tools/generate-templates.sh
Comment (apatard) [reply tools/generate-templates.sh:9]: I'm not sure what's the best way to do it. If it has to be hardcoded, so be it. Maybe try to do as you said: look at the src/molecule_plugins content ?

File: tools/generate-templates.sh
Comment (wadhah101) [reply tools/generate-templates.sh:6]: Some errors are unfixable. And the fixable ones are too much to be included in this PR. there's no going around `.ansible-lint-ignore` file

File: tools/generate-templates.sh
Comment (wadhah101) [reply tools/generate-templates.sh:9]: I will try to find a way to grab the driver names from `pyproject.toml`

```toml
[project.entry-points."molecule.driver"]
azure = "molecule_plugins.azure.driver:Azure"
containers = "molecule_plugins.containers.driver:Container"
docker = "molecule_plugins.docker.driver:Docker"
ec2 = "molecule_plugins.ec2.driver:EC2"
gce = "molecule_plugins.gce.driver:GCE"
podman = "molecule_plugins.podman.driver:Podman"
vagrant = "molecule_plugins.vagrant.driver:Vagrant"
```

File: tools/generate-templates.sh
Comment (zhan9san) [reply tools/generate-templates.sh:6]: It's fun to fix the real lint issue in another issue.

Ignoring them to make the PR smaller is acceptable.

File: tools/generate-templates.sh
Comment (wadhah101) [reply tools/generate-templates.sh]: @apatard I have generated the package names from `pyproject.toml`, I think the code looks ok now. Can you check ?

File: tools/generate-templates.sh
Comment (apatard) [inline tools/generate-templates.sh:8]: why not making the python script outputting the list of drivers ?
This would avoid using jq and if we need later a way to get the list of drivers, the python script would be very handy.

Comment (wadhah101) [reply tools/generate-templates.sh]: @apatard Yeah at first i wanted to do it without a script with some shell command but couldn't find any reliable python-cli toml parser.
I have updated it to output the plugins, it makes more sense this way

**Comment by apatard** [comment]
I guess this PR is near to be ready to be merged. I hope that once merged, we'll manage to reduce the ignore list.

File: requirements.yml
Comment (wadhah101) [reply requirements.yml:1]: ansible lint will complain if `azure.azcollection` is not installed.

for `crypto` the version is defined in `src/molecule_plugins/gce/driver.py` so it's a good idea to fix it too

Comment (apatard) [reply requirements.yml:1]: the crypto change seems weird/wrong. I mean 1.8.0 is old and we should not install old stuff. I don't know why it's saying 1.8.0 but I guess it should rather be >= 1.8.0...

Comment (wadhah101) [reply requirements.yml:1]: done

**Comment by apatard** [comment]
@zhan9san while we're waiting for github actions to be working again, do you have remaining comments ?

**Review by zhan9san — APPROVED** [review]

**Comment by zhan9san** [comment]
LGTM.

Is Github Action down?

**Comment by wadhah101** [comment]
Github actions seem to be back. Can any of the maintainers re run the checks ?

**Comment by apatard** [comment]
Things are not perfect but the CI is green, so I guess it's time to merge. The side-effects (if any) can be fixed later. It's better than not having -lint on the cookie cutter stuff. Thanks

**Pull request merged**
